warehouse/internal/repo/schema.go

```diff
@@ -105,22 +108,137 @@ func (sh *WHSchema) Insert(ctx context.Context, whSchema *model.WHSchema) (int64
+				}
+				err = sh.insertTableSchema(ctx, tx, tableWHSchema, tableSchemaPayload)
+				if err != nil {
+					return fmt.Errorf("inserting table-level schema for table %s: %w", tableName, err)
+				}
+			}
+		}
+
 		return nil
 	})
 
-	return id, err
+	return err
 }
 
-func (sh *WHSchema) GetForNamespace(ctx context.Context, destID, namespace string) (model.WHSchema, error) {
+// updateOtherTableLevelSchemas updates all other schemas with the same destination_id, namespace, and table_name but different source_id.
+func (sh *WHSchema) updateOtherTableLevelSchemas(ctx context.Context, tx *sqlmiddleware.Tx, whSchema *model.WHSchema, tableName string, tableSchemaPayload []byte, now time.Time) error {
+	_, err := tx.ExecContext(ctx, `
+		UPDATE `+whSchemaTableName+`
+		SET
+			schema = $1,
+			updated_at = $2,
+			expires_at = $3
+		WHERE
+			destination_id = $4 AND
+			namespace = $5 AND
+			source_id != $6 AND
+			table_name = $7;
+	`,
+		tableSchemaPayload,
+		now.UTC(),
+		whSchema.ExpiresAt,
+		whSchema.DestinationID,
+		whSchema.Namespace,
+		whSchema.SourceID,
+		tableName,
+	)
+	if err != nil {
+		return fmt.Errorf("updating other table-level schemas for table %s: %w", tableName, err)
+	}
+	return nil
+}
+
+// insertTableSchema inserts or updates a table-level schema row in wh_schemas with the given table name, using the provided transaction and marshaled schema.
+func (sh *WHSchema) insertTableSchema(ctx context.Context, tx *sqlmiddleware.Tx, tableWHSchema *model.WHTableSchema, tableSchemaPayload []byte) error {
+	_, err := tx.ExecContext(ctx, `
+		INSERT INTO `+whSchemaTableName+` (
+		  source_id, namespace, destination_id,
+		  destination_type, schema, created_at,
+		  updated_at, expires_at, table_name
+		)
+		VALUES
+		  ($1, $2, $3, $4, $5, $6, $7, $8, $9)
+		ON CONFLICT (
+			source_id, destination_id, namespace, table_name
+		) DO
+		UPDATE
+		SET
+		  schema = $5,
+		  updated_at = $7,
+		  expires_at = $8;
+	`,
+		tableWHSchema.SourceID,
+		tableWHSchema.Namespace,
+		tableWHSchema.DestinationID,
+		tableWHSchema.DestinationType,
+		tableSchemaPayload,
+		tableWHSchema.CreatedAt.UTC(),
+		tableWHSchema.UpdatedAt.UTC(),
+		tableWHSchema.ExpiresAt,
+		tableWHSchema.TableName,
+	)
+	if err != nil {
+		return fmt.Errorf("inserting table-level schema: %w", err)
+	}
+	return nil
+}
+
+// GetForNamespace fetches the schema for a namespace, supporting both legacy and table-level modes.
+func (sh *WHSchema) GetForNamespace(ctx context.Context, destID, namespace string, enableTableLevelSchema bool) (model.WHSchema, error) {
+	if !enableTableLevelSchema {
+		return sh.getForNamespace(ctx, destID, namespace)
+	}
+
+	originalSchema, err := sh.getForNamespace(ctx, destID, namespace)
+	if err != nil {
+		return model.WHSchema{}, err
+	}
+
+	tableLevelSchemas, err := sh.getTableLevelSchemasForNamespace(ctx, destID, namespace)
+	if err != nil {
+		return model.WHSchema{}, err
+	}
+	if !reflect.DeepEqual(originalSchema.Schema, tableLevelSchemas) {
```

Comment:
I'm assuming that this assertion will go away once we are confident with the table level code ?